code/datums/mapgen/CaveGenerator.dm

```diff
@@ -106,7 +125,95 @@
 	to_chat(world, span_boldannounce("[message]"))
 	log_world(message)
 
+
+/**
+ * This proc handles including biomes in the cave generation. This is slower than
+ * `generate_terrain()`, so please use it only if you actually need biomes.
+ *
+ * This should only be called by `generate_terrain()`, if you have to call this,
+ * you're probably doing something wrong.
+ */
+/datum/map_generator/cave_generator/proc/generate_terrain_with_biomes(list/turfs, area/generate_in)
+	if(!(generate_in.area_flags & CAVES_ALLOWED))
+		return
+
+	var/humidity_seed = rand(0, 50000)
+	var/heat_seed = rand(0, 50000)
+
+	var/start_time = REALTIMEOFDAY
+	string_gen = rustg_cnoise_generate("[initial_closed_chance]", "[smoothing_iterations]", "[birth_limit]", "[death_limit]", "[world.maxx]", "[world.maxy]") //Generate the raw CA data
+
+	var/list/open_turfs_used = list()
+
+	for(var/turf/gen_turf as anything in turfs) //Go through all the turfs and generate them
+		var/closed = string_gen[world.maxx * (gen_turf.y - 1) + gen_turf.x] != "0"
+		var/turf/new_turf = pick(closed ? closed_turf_types : open_turf_types)
+
+		var/datum/biome/selected_biome
+
+		// Here comes the meat of the biome code.
+		var/drift_x = (gen_turf.x + rand(-BIOME_RANDOM_SQUARE_DRIFT, BIOME_RANDOM_SQUARE_DRIFT)) / perlin_zoom
+		var/drift_y = (gen_turf.y + rand(-BIOME_RANDOM_SQUARE_DRIFT, BIOME_RANDOM_SQUARE_DRIFT)) / perlin_zoom
+
+		var/humidity = text2num(rustg_noise_get_at_coordinates("[humidity_seed]", "[drift_x]", "[drift_y]"))
+		var/heat = text2num(rustg_noise_get_at_coordinates("[heat_seed]", "[drift_x]", "[drift_y]"))
+		var/heat_level //Type of heat zone we're in (LOW-LOWMEDIUM-HIGHMEDIUM-HIGH)
+		var/humidity_level  //Type of humidity zone we're in (LOW-LOWMEDIUM-HIGHMEDIUM-HIGH)
+
+		switch(heat)
+			if(0 to 0.25)
+				heat_level = BIOME_LOW_HEAT
+			if(0.25 to 0.5)
+				heat_level = BIOME_LOWMEDIUM_HEAT
+			if(0.5 to 0.75)
+				heat_level = BIOME_HIGHMEDIUM_HEAT
+			if(0.75 to 1)
+				heat_level = BIOME_HIGH_HEAT
+
+		switch(humidity)
+			if(0 to 0.25)
+				humidity_level = BIOME_LOW_HUMIDITY
+			if(0.25 to 0.5)
+				humidity_level = BIOME_LOWMEDIUM_HUMIDITY
+			if(0.5 to 0.75)
+				humidity_level = BIOME_HIGHMEDIUM_HUMIDITY
+			if(0.75 to 1)
+				humidity_level = BIOME_HIGH_HUMIDITY
+
+		selected_biome = possible_biomes[heat_level][humidity_level]
+		selected_biome = SSmapping.biomes[selected_biome] //Get the instance of this biome from SSmapping
+
+		// Currently, we only affect open turfs, because biomes don't currently
+		// have a definition for biome-specific closed turfs.
+		if((!length(turfs_affected_by_biome) && !closed) || turfs_affected_by_biome[new_turf])
```

Comment:
turfs_affected_by_biome is totally unused shouldn't be here

Reply:
`turfs_affected_by_biome` is there to be a whitelist of which turfs _can_ actually be replaced. An example of this would be, let's say we want to keep the ice in ice box there, but we want to swap the snow for, I don't know, a different kind of snow (like, packed snow, or fluffier snow, or whatever). This allows the ice turfs to be skipped, whereas the rest will be replaced.

In my usecase, it was to replace a generic type of turf with its specific biome variant, whilst there was some dirt that I didn't want to get replaced.

Reply:
In my test case, I also used it to generate something only on the non-ice tiles of the icebox caves, just to prove that it works.

Reply:
I don't understand why you'd do this as an assoc list instead of as an overridable proc that's called on every turf if a variable on the biome or or something is set.

when is this list meant to be populated?